bokeh/state.py

```diff
@@ -0,0 +1,568 @@
+
+        title (str, optional) : a title for the HTML document (default: None)
+            If None, use the default state title value, if there is one.
+            Otherwise, use "Bokeh Plot"
+
+    Returns:
+        None
+
+    Raises:
+        RuntimeError
+
+    """
+    if state is None:
+        state = _state
+
+    filename, resources, title = _get_save_args(state, filename, resources, title)
+
+    _save_helper(obj, filename, resources, title)
+
+def _get_save_args(state, filename, resources, title):
+
+    if filename is None and state.file:
+        filename = state.file['filename']
+
+    if resources is None and state.file:
+        resources = state.file['resources']
+
+    if title is None and state.file:
+        title = state.file['title']
+
+    if filename is None:
+        raise RuntimeError("save() called but no filename was supplied and output_file(...) was never called, nothing saved")
+
+    if resources is None:
+        warnings.warn("save() called but no resources was supplied and output_file(...) was never called, defaulting to resources.INLINE")
+        from .resources import INLINE
+        resources = INLINE
+
+    if title is None:
+        warnings.warn("save() called but no title was supplied and output_file(...) was never called, using default title 'Bokeh Plot'")
+        title = "Bokeh Plot"
+
+    return filename, resources, title
+
+def _save_helper(obj, filename, resources, title):
+
+    # TODO: (bev) Widget seems awkward as a base class to check here
+    if isinstance(obj, Widget):
+        doc = Document()
+        doc.add(obj)
+
+    elif isinstance(obj, Document):
+        doc = obj
+
+    else:
+        raise RuntimeError("Unable to save object of type '%s'" % type(obj))
+
+    html = file_html(doc, resources, title)
+    with io.open(filename, "w", encoding="utf-8") as f:
+        f.write(decode_utf8(html))
+
+def push(session=None, document=None, state=None):
+    """ Update the server with the data for the current document.
+
+    Will fall back to the default output state (or an explicitly provided
+    :class:`State` object) for ``session`` or ``document`` if they are not
+    provided.
+
+    Args:
+        session (Session, optional) : a Bokeh server session to push objects to
+
+        document (Document, optional) : A :class:`bokeh.document.Document` to use
+
+    Returns:
+        None
+
+    """
+    if state is None:
+        state = _state
+
+    if not session:
+        session = state.session
+
+    if not document:
+        document = state.document
+
+    if not session:
+        warnings.warn("push() called but no session was supplied and output_server(...) was never called, nothing pushed")
+        return
+
+    return session.store_document(document)
+
+def reset_output(state=None):
+    ''' Clear the default state of all output modes.
+
+    Returns:
+        None
+
+    '''
+    _state.reset()
+
+def _deduplicate_plots(plot, subplots):
```

Comment:
Quite possibly. I really want to get @mattpap opinion on these. It might make sense to move all the functions, and the global `_state` plus some things like `grid`, `vbox` and `hbox` to a shared module as you describe below.